### ISSUE TITLE
`int` is not iterable, when variable cannot be an int

### ISSUE THREAD

**Issue by EvgeniGenchev** [microsoft/pyright#4793]
**Describe the bug**
I have this class function that returns either int or a list:
![image](https://user-images.githubusercontent.com/59848681/225882348-dbd7d3e5-cf63-41d4-bd22-a877eea8ab12.png)

I import it into another file and if the result is of a type int I return a message if not I want to interate over the variable like so:

![image](https://user-images.githubusercontent.com/59848681/225882652-b1bb0568-f0e8-4ffe-b029-7b3e87b22cce.png)

However, as you can see `query` cannot be `int` and can be only a `list`
 
**To Reproduce**
Well create a file with an arbiterary name like so:

```python
class A:
    # some function that can return either int or list
    def function(self, i:int) -> int | list:
        if i % 2 == 0:
            return 1
        else:
            return []


def other_function(i):
    object = A()

    result = object.function(i)

    if type(result) is int:
        return 0
    
    # "int" is not iterable  "__iter__" method not defined
    for _i in result: 
        print(_i)
```
Then run `pyright <file_name>` and it will complain that the `"int" is not iterable`

**Expected behavior**
The code above works fine, since the for loop block is unreachable if result is an `int`, thus `pyright`
is expected to not indentify this as an error. 

**Additional context**
Test both on neovim thought the LSP and running by running the `pyright <file_name>` command from the terminal.

<br>Versions:

```sh
pyright 1.1.295
NVIM v0.8.3
```

**Comment by erictraut** [comment]
In general, the type guard for the pattern `type(x) is y` cannot narrow the type in the negative ("else") code path unless the types in the union are marked `@final`. In this particular case, I think it would be safe to make this assumption. I'll look at improving the logic in the negative narrowing case.

In the meantime, if you switch to an `isinstance(result, int)` check, negative narrowing will work as you expect.

**Comment by EvgeniGenchev** [comment]
My bad in the "to reproduce code I used `==`, however, in the original code (see screenshots) where the bug was caught I used `is`. And running:

```python
class m(int):
    pass

print(m() is int)
``` 

Shows that I class inheriting `int` is not an `int` when using the `is` operator. So the bug is real but I made a mistake in reproducing it, sorry...

**Comment by EvgeniGenchev** [comment]
@erictraut Then the question shifts to if my typing specifies it returns an `int`, why me returning a class inheriting from `int` is not flagged by pyright? Very poor but simple exmple of why that would be a problem is the following:

```python
class MyInt(int):
    def __div__(self, other):
        if self == 0:
            return 0
        return self / other

    def __add__(self, other):
        return self + other + 1

class A:
    # some function that can return either int or list
    def function(self, i:int) -> int | list:
        if i % 2 == 0:
            return MyInt(1)
        else:
            return []


def other_function(i):
    object = A()

    result = object.function(i)

    if type(result) is int:
        return 0
    
    for _i in result: #pyright: ignore
        print(_i)

```

**Comment by erictraut** [comment]
This isn't specific to pyright. The Python type system (as defined in PEP 484) specifies that a type annotation of `int` indicates that a compatible type must be an instance of `int` _or a subclass thereof_. Since `MyInt` is a subclass of `int`, it is compatible with a return type of `int`. 

Without this subclass provision, you would receive errors for something as simple as `def foo() -> float: return 0` because `0` is an `int`, not a `float`.

The assumption in object-oriented programming is that a subclass must honor the interface of its superclass. Of course, you can violate this assumption if you try hard enough (as in your example above), but that would be considered a programming error in OOP. Type theory folks give this the name [Liskov Substitution Principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle).

**Comment by EvgeniGenchev** [comment]
Gotta love the type theory folks. Thanks for looking into the issue and sorry for wasting your time!
